internal/provider/job_resource.go

```diff
@@ -0,0 +1,518 @@
+		if err != nil {
+			resp.Diagnostics.AddError(
+				"Unexpected Resource Read error",
+				err.Error(),
+			)
+			return
+		}
+
+		if http_code != http.StatusOK {
+			resp.Diagnostics.AddError(
+				"Unexpected Resource Read error",
+				fmt.Sprintf("The server returned status code %d while attempting to Get from URL %s", http_code, data.URL.ValueString()),
+			)
+			return
+		}
+
+		tflog.Info(ctx, fmt.Sprintf("HTTP GET [%s] => %s", data.URL.ValueString(), string(body)))
+		err = data.ParseHttpResponse(body)
+		if err != nil {
+			resp.Diagnostics.AddError(
+				"Unexpected Resource Read error",
+				"Failed to parse HTTP JSON Response: "+err.Error(),
+			)
+			return
+		}
+	}
+
+	// Save updated data into Terraform state
+	resp.Diagnostics.Append(resp.State.Set(ctx, &data)...)
+}
+
+func (r JobResource) Create(ctx context.Context, req resource.CreateRequest, resp *resource.CreateResponse) {
+	var data *jobResourceModel
+
+	// Read Terraform plan data into the model
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &data)...)
+
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// Create new Job from job template
+	req_data, err := data.CreateRequestBody(ctx)
+	if err != nil {
+		resp.Diagnostics.AddError(
+			"Unexpected Resource Creation error",
+			err.Error(),
+		)
+	}
+	var http_code int
+	var body []byte
+	post_url := "/api/v2/job_templates/" + data.Id.String() + "/launch/"
+	if req_data != nil {
+		http_code, body, err = r.client.doRequestWithBody("POST", post_url, req_data)
+	} else {
+		http_code, body, err = r.client.doRequest("POST", post_url)
+	}
+
+	if err != nil {
+		resp.Diagnostics.AddError(
+			"Unexpected Resource Creation error",
+			err.Error(),
+		)
+		return
+	}
+
+	if http_code != http.StatusCreated {
+		resp.Diagnostics.AddError(
+			"Unexpected Resource Creation error",
+			fmt.Sprintf("The server returned status code %d while attempting to create Job", http_code),
+		)
+		return
+	}
+
+	tflog.Info(ctx, fmt.Sprintf("HTTP POST [%s] => %s", post_url, string(body)))
+	err = data.ParseHttpResponse(body)
+	if err != nil {
+		resp.Diagnostics.AddError(
+			"Unexpected Resource Creation error",
+			"Failed to parse HTTP JSON Response: "+err.Error(),
+		)
+		return
+	}
+
+	if data.WaitForCompletion.ValueBool() {
+		var wait_duration int64 = 120
+		if !(data.WaitDuration.IsNull() && data.WaitDuration.IsUnknown()) {
+			wait_duration = data.WaitDuration.ValueInt64()
+		}
+		job_status, err := r.WaitForJob(data.URL.ValueString(), []string{"successful", "failed"}, wait_duration)
+		if err != nil {
+			resp.Diagnostics.AddError(
+				"Unexpected Resource Creation error",
+				"An error occurred while waiting for Job to complete: "+err.Error(),
+			)
+			return
+		}
+		data.Status = types.StringValue(job_status)
+	}
+
+	// Save updated data into Terraform state
+	resp.Diagnostics.Append(resp.State.Set(ctx, &data)...)
+}
+
+func (r JobResource) Delete(ctx context.Context, req resource.DeleteRequest, resp *resource.DeleteResponse) {
```

Comment:
I don't think we should be deleting jobs by default; many users may want to keep their job launch history to be able audit it.  The user running the job may not even have access to delete things.  What prompted the decision to delete jobs in Controller?

Reply:
All four CRUD operations are [expected for resources](https://developer.hashicorp.com/terraform/plugin/framework/resources) in the terraform provider framework. I was chatting about this with Mike the other day, but IMO launching an AAP job is similar to invoking a lambda function in AWS and I'm wondering if we want our provider to handle it in a similar way? I notice the AWS provider allows you to invoke a lambda either as [a resource](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/lambda_invocation) or as [a data source](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/data-sources/lambda_invocation), which is interesting.

Reply:
@jillr as mentioned by @hakbailey we need all CRUD operation to be defined for a resource. Now we can discuss about the delete operation whether it should simply cancel the job or fully delete it. Also note that if the Job was launched by the user they should be able to delete it too.

Reply:
While the CRUD operations have to be there for the Terraform data model, delete and update don't actually have to do anything on the AAP side. I'm not sure that they should, either. Our goal isn't to build a Terraform resource that fully implements the AAP API, our goal is to allow Terraform users to connect changes in an inventory to launching a job template. This is really the primary use case. I think we want to emphasize the lifecycle connection between Instance --> Inventory --> Job.

As @hakbailey suggested, the way the AWS provider handles the lambda invocation is probably a good model to look at. I'm not sure that using a data source is the right approach as this would launch a job template for every apply. Using a resource gives the user the flexibility to decide whether the job template should be launched only on create, or when some other arbitrary property has changed. For example, you might add a trigger that looks at the contents of an inventory, so the job template gets launched whenever hosts are added/removed from the inventory.

Reply:
Thanks y'all.  I think we should avoid deleting jobs on the AAP side as I expect a lot of AAp users would be surprised by that.

Reply:
I just want to summarize for everyone to be aligned, 

***Current implementation***

- _Update_: a new job will be triggered if one of the following has changed: job_template_id, ee_id, inventory_id or extra_vars, this is internal TF logic
- _Delete_: cancel running job and delete it from AAP 

***What we are expecting***

- _Update_: ? I am not clear about what you are expecting
- _Delete_: Cancel running Job if not finished

@jillr @gravesm

Reply:
When an update is required, launch a new job. An update should be triggered when the inventory_id, job_template_id or extra_vars are changed. I would remove the ee_id and wait properties from the resource. Our goal, for now at least, should be to provide a minimally useful mechanism for spawning a job template. The expectation is still that the user is going to be interacting with the AAP UI to configure job templates and check the status of jobs.

The resource should also have a trigger property that accepts an arbitrary set of key/values. This can be used to trigger an update of the resource based on whatever criteria the user wants.

The resource should not cancel or delete an AAP job at all. Deleting just removes the resource from the TF state.